Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt

```diff
@@ -262,9 +285,51 @@ The canonical form of a curve is obtained by flattening its brackets, removing i
+\cgalFigureBegin{fig_relative_order_normal,relative_order_normal.svg}
+ The red edge is being processed and is compared against the pink edge. Since the green edge (the predecessor of the red edge) is to the right of the blue edge around the vertex, red edge must be to the right of the pink edge in the ordering.
+    \cgalFigureEnd
+
+-# If the current edge crosses the same edge in the quadrangulation as the first edge of the path, the predecessor of the first edge (aka the last edge of the path) has not been processed yet. Hence if they form the same turn, we cannot determine the relative order of the predecessors. So the idea is to keep following the predecessors of current edge and the first edge until they diverge, at which point the relative order around the vertex can be used to determine the relative order. See \cgalFigureRef{fig_relative_order_first}. This can be precomputed by finding the all the longest common suffices of the path against its circular shifts. A modified Knuth-Morris-Pratt algorithm is applied to preprocess the path in linear time.
+\cgalFigureBegin{fig_relative_order_first,relative_order_first.svg}
+ The red edge is being processed and is compared against the pink edge which is the first edge of the path. The blue and green edges are the first diverging pair when tracing backward. The dashed line means that edges may not be processed yet. Since the green edge is to the right of the blue edge around the vertex, red edge must be to the right of the pink edge in the ordering.
+    \cgalFigureEnd
+
+The ordering is stored in a red-black tree on each edge of the quadrangulation. So each insertion and search takes \f$O(\log{l})\f$ time.
+
+\subsubsection SMTopology_Simplicity_Test_Verification Verify Ordering
+After computing a tentative ordering within the edges of the path, we have to verify such ordering could result in a intersection free arrangement. Since there is no intersection within an edge, we only need to verify this for each vertex in the quadrangulation. We can consider the two consecutive edges composing a turn (one going in the vertex, one going out of the vertex) at the vertex being verified a <em>pair</em>. The ordering at the vertex is intersection free if and only if there is no two pairs crossing each other according to the clockwise ordering around the vertex. In other words, for any two pairs \f$(a, a')\f$ and \f$(b, b')\f$, subsequence \f$a, b, a', b'\f$ should not appear in the clockwise ordering. This is very similar to verifying balanced parentheses in a string. We traverse clockwise at each vertex and use a stack-based algorithm to verify in linear time that the ordering produces a cycle without self-intersection.
```

Comment:
```suggestion
After computing a tentative ordering within the edges of the path, we have to verify that such an ordering could result in an intersection free arrangement. Since there is no intersection within an edge, we only need to verify this for each vertex in the quadrangulation. We can consider the two consecutive edges composing a turn (one going in the vertex, one going out of the vertex) at the vertex being verified a <em>pair</em>. The ordering at the vertex is intersection free if and only if there is no two pairs crossing each other according to the clockwise ordering around the vertex. In other words, for any two pairs \f$(a, a')\f$ and \f$(b, b')\f$, subsequence \f$a, b, a', b'\f$ should not appear in the clockwise ordering. This is very similar to verifying balanced parentheses in a string. We traverse clockwise at each vertex and use a stack-based algorithm to verify in linear time that the ordering produces a cycle without self-intersection.
```